### PR TITLE
Advanced Building Blocks - Bubble Sort 

### PR DESCRIPTION
**Pull request by Victorgonbu**
# Added

- Methods bubble_sort and bubble_sort_by.
- Input for arrays.
- Outputs an array sorted by number of characters  and an array sorted by number.

### CODE REVIEW

**Comment by addod19** [comment]
## Pull Request under review by Daniel Larbi Addo (TSE)

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @Shaqri and @sipe-daniel, great job on this project :+1: PR is quite descriptive, great commits messages :slightly_smiling_face: Let us improve our project by fixing the comments below:

1. Kindly create a new file for the ```bubble_sort_by``` method
2. Kindly implement the use of ```yield``` and ```passing a block``` to a method
3. Kindly improve the ```sort_by_method```
4. Kindly improve the ```README``` file

Kindly fix these comments and send another code review :+1: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Comment by sipe-daniel** [comment]
Thanks for your comments about our bubble project.
Please, we are confused about your second and third remarks. Yield is not a requirement for the first method and your third comment is not clear.


#Misunderstanding

The remarks:
2. yield is not a requirement about the first method
3. How can we improve sort_by_method?

File: bubble_sort_by.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      if yield(an_array[j - 1], an_array[j]).positive?
+        an_array[j - 1], an_array[j] = an_array[j], an_array[j - 1]
+        counter += 1
+      end
+      j += 1
+    end
+    counter.zero? ? break : i += 1
+  end
+  an_array
+end
```
Comment (christianotieno) [inline bubble_sort_by.rb:17]: Tests Results:

Annotation 2020-07-10 045005

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey @Shaqri  & @sipe-daniel :wave:
Good job so far! :tada: We have some few changes to do before we proceed (see review comment).

If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.*** :eyes:
- First, to clarify on the first review. We have an issue in the `bubble_sort_by` method.
- [x] During test runs, the method should sort an array by accepting a block and raise an error if a block is not given. Apparently this doesn't seem to be the case. Let's work on this. Refer below to the failed test and improve.
- [x] There are some grammatical errors in the last message being printed to the user. Let;'s work on that as well.

- When you have made the specified changes, please submit another code review. Thank you!
*Happy coding :smile:,*
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Comment by Victorgonbu** [comment]
Thank you very much for reviewing our code Christian. We made the changes requested

**Comment by addod19** [comment]
## Pull Request under review by Daniel Larbi Addo (TSE)

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @Shaqri and @sipe-daniel, great job implementing the changes :+1: However, there is a bug we need to fix. Kindly check comments below: :slightly_smiling_face: 
- ```First when I run ruby bubble_sort_by.rb``` I get this error
ER
- Then I replaced ```length``` with ```size``` in the block and run ```bubble_sort_by.rb```. The output was not sorted
pd

Kindly fix these and submit another code review :+1: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](TSE)

**Comment by Shaqri** [comment]
Hi @addod19, i have checked my code and it's working perfectly. By the screenshot second you sent us, i can tell it's not our bubble_sort project. Please let me know if i'm wrong.
There is a link to repl.it where out project can be tested.
Than you very much for reviewing!

**Comment by pelzolga123** [comment]
## Hi Team, 
### You did a good job so far! Still, there are more changes required:
* [x] `bubble_sort_by` should raise an error if block is not given. For example:
When trying to test method with this piece of code: `p bubble_sort_by(['hi', 'hey', 'hello'])` you should get next error:  
```ruby
Traceback (most recent call last):
        5: from enum.rb:31:in `<main>'
        4: from enum.rb:21:in `bubble_sort_by'
        3: from enum.rb:21:in `times'
        2: from enum.rb:22:in `block in bubble_sort_by'
        1: from enum.rb:22:in `times'
enum.rb:23:in `block (2 levels) in bubble_sort_by': no block given (yield) (LocalJumpError)
```
* [x] No debugging instructions should be in committed code (as discussed [here](https://questions.microverse.org/t/debugging-instructions-code/346)) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/section/debugging_requirement.png) from Pathwright). This means that you should delete everything except for the method. All test cases you have there right now should be removed.

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

**Comment by Victorgonbu** [comment]
Thank you for reviewing @pelzolga123.
Made requested changes

**Comment by lekegitrepo** [comment]
### Hi Team :clap: :clap:
### Good job so far. As much as I would have liked to approve this project, however, your implementation did not reflect the last review instruction about the bubble_sort_by. As mentioned by the last reviewer the method should raise an error if block is not given. When I ran your code it outputs "Block not given" instead of raising an error. Please fix this and submit for another review. You are almost there :muscle: .

##### _Reviewed by Temitayo Adeleke_.

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
##### Status: Request changes

**Review by mikenath223 — APPROVED** [review]
### CONGRATULATIONS  👏👏👏

This project meets the requirements for approval.

**STATUS: APPROVED ✅**


Don't forget to fill the project completion form👍
Good luck with the next project.

HAPPY CODING👨‍💻

**Pull request merged**